### PR TITLE
Fix diagonal 0 error in row-reduction process

### PR DESCRIPTION
**Pull request by DeqingSun**
This fix https://github.com/v923z/micropython-ulab/issues/173
And the accuracy was pretty good compare to numpy.

### CODE REVIEW

File: code/linalg/linalg.c [hunk 1]
```diff
@@ -157,10 +174,27 @@ static mp_obj_t linalg_det(mp_obj_t oin) {
         tmp[i] = ndarray_get_float_value(in->array->items, in->array->typecode, i);
     }
     mp_float_t c;
+    mp_float_t det_sign = 1.0;
```
Comment (v923z) [inline code/linalg/linalg.c:177]: On some platforms, `mp_float_t det_sign = MICROPY_FLOAT_CONST(1.0)` is required. Could you, please, replace it?

File: code/linalg/linalg.c [hunk 2]
```diff
@@ -55,8 +55,25 @@ bool linalg_invert_matrix(mp_float_t *data, size_t N) {
     for(size_t m=0; m < N; m++){
         // this could be faster with ((c < epsilon) && (c > -epsilon))
         if(MICROPY_FLOAT_C_FUN(fabs)(data[m*(N+1)]) < epsilon) {
-            m_del(mp_float_t, unit, N*N);
-            return false;
+            //look for a line to swap
```
Comment (v923z) [inline code/linalg/linalg.c:58]: Add your name on line 12!

**Comment by v923z** [comment]
@DeqingSun Many thanks for looking into this! This could be merged, after you fix the two small issues I indicated in the comments.

File: code/linalg/linalg.c (hunk 1 above)
Comment (DeqingSun) [reply code/linalg/linalg.c:177]: Will replace this one. I didn't know such Marco. After quick search it seems multiple files need to be changed as well?

```
$ grep -Er ' [0-9]+\.' .
./linalg/linalg.c:    mp_float_t elem = 1.0;
./linalg/linalg.c:	mp_float_t sum = 0.0;
./linalg/linalg.c:    mp_float_t det_sign = 1.0;
./linalg/linalg.c:        mp_float_t dot = 0.0;
./linalg/linalg.c:                mp_float_t sum = 0.0, v1, v2;
./linalg/linalg.c:        eigvectors[m*(in->n+1)] = 1.0;
./linalg/linalg.c:        largest = 0.0;
./linalg/linalg.c:    mp_float_t dot = 0.0, v;
./linalg/linalg.c:	mp_float_t trace = 0.0;
./ulab_create.c://|       Difference between consecutive elements, optional, defaults to 1.0
./ulab_create.c:        start = 0.0;
./ulab_create.c:        step = 1.0;
./ulab_create.c:        step = 1.0;
./ulab_create.c://|        and the other elements set to 0."""
./ulab_create.c://|    Return a new array of the given shape with all elements set to 1."""
./ulab_create.c://|    Return a new array of the given shape with all elements set to 0."""
./numerical/numerical.c:    mp_float_t value, sum = 0.0, sq_sum = 0.0;
./numerical/numerical.c:        sum = sq_sum = 0.0;
./numerical/numerical.c:        sum = 0.0;
./numerical/numerical.c:        sum_sq = 0.0;
./numerical/numerical.c:    // TODO: transpose the matrix, if axis == 0. Though, that is hard on the RAM...
./approx/approx.c://|     xtol: float = 2.4e-7,
./approx/approx.c://|     xatol: float = 2.4e-7,
./approx/approx.c://|     fatol: float = 2.4e-7,
./approx/approx.c:    //mp_float_t lambda = 0.0078125;
./approx/approx.c://|     xtol: float = 2.4e-7,
./approx/approx.c://|     rtol: float = 0.0,
./approx/approx.c://| def trapz(y: ulab.array, x: Optional[ulab.array] = None, dx: float = 1.0) -> float:
./ulab.h:// poly requires approx. 2.5 kB
./ulab.h:// the approx module consumes about 4.5 kB of flash space
./poly/poly.c:        XT[i+0*lenx] = 1.0; // top row
./poly/poly.c:            sum = 0.0;
./poly/poly.c:        sum = 0.0;
./poly/poly.c:        sum = 0.0;

```

Comment (DeqingSun) [reply code/linalg/linalg.c]: Also, is it necessary to change ```mp_obj_new_float(0.0);```? I guess so?

Comment (v923z) [reply code/linalg/linalg.c]: Yeah, that's correct.

**Comment by v923z** [comment]
@jonathanhogg Jonathan, I was wondering, whether you could comment on this PR. You submitted https://github.com/v923z/micropython-ulab/pull/154, which addressed the problem of float constants. Deqing found a couple of other occurrences (see his long comment above), and I am not sure, how they slipped through last time. Since I don't see the problem that your PR fixed, I am a bit clueless, and could use a bit of advice here. Thanks!

**Comment by jonathanhogg** [comment]
My understanding is that if you have code along the lines of:

```C
mp_float_t foo = 1.0;
```

then the compiler will be happy. On platforms where `mp_float_t` resolves to `float`, it will infer that `1.0` needs to be a `float` constant. However, if you have code like:

```C
mp_float_t foo;
foo = bar * 3.457;
```

It will take `3.457` as a `double` constant, compile a double-precision multiplication with `bar` and then cast it to a `float` for assignment to `foo`. This generates a compiler warning and, on the ESP32 port at least, the default build arguments interpret all warnings as errors and so the build fails.

So it's just the constants that are used in expressions that have to be wrapped in `MICROPY_FLOAT_CONST()`. There is, of course, no harm in just wrapping all float constants…

Either way, it'd be good to add an ESP32 platform build to the continuous-integration checks.

**Comment by v923z** [comment]
> So it's just the constants that are used in expressions that have to be wrapped in `MICROPY_FLOAT_CONST()`. There is, of course, no harm in just wrapping all float constants…

Thanks for the thorough explanation!

> 
> Either way, it'd be good to add an ESP32 platform build to the continuous-integration checks.

I am not against the idea, but I don't work with the ESP32, so it would be a bit of hassle for me. However, if you are willing to update the workflow file, I would be happy to merge that.

File: code/linalg/linalg.c (hunk 1 above)
Comment (DeqingSun) [reply code/linalg/linalg.c]: MICROPY_FLOAT_CONST added. At least linalg.c is good to go.

**Comment by jonathanhogg** [comment]
> I am not against the idea, but I don't work with the ESP32, so it would be a bit of hassle for me. However, if you are willing to update the workflow file, I would be happy to merge that.

Figured you might say that 😉 – I'll add it to my to-do list…

**Comment by v923z** [comment]
> Figured you might say that wink – I'll add it to my to-do list…

Thanks!

File: code/linalg/linalg.c (hunk 1 above)
Comment (v923z) [reply code/linalg/linalg.c]: Thanks again for seeing this through till the end! You got your brownie points: https://github.com/v923z/micropython-ulab/graphs/contributors.

**Pull request merged**